### PR TITLE
[DPE-2992] Part-1 Recieve keyfile from config-server

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
DPE-2992 is to start the mongos subordinate charm with the provided config server. In order for the config server and for mongos to be connected they both must have the same keyfile

## Solution
Recieve the keyfile

## Follow up PRs
2. Updating the library to actually start the mongos service with the information of the config server
3. Integration tests

## Testing
```
juju deploy ./*charm --config role="config-server"
juju deploy application
juju deploy mongos
juju integrate application mongos
juju integrate mongos:cluster mongodb:cluster
juju ssh mongos
sudo cat KEY_FILE_PATH
```

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -90,6 +105,151 @@ def _install_snap_packages(self, packages: List[str]) -> None:
...
+
+        # MongoDB limitation; it is needed 400 rights for keyfile and we need 440 rights on tls
+        # certs to be able to connect via MongoDB shell
+        if Config.TLS.KEY_FILE_NAME in file_name:
+            os.chmod(file_name, 0o400)
+        else:
+            os.chmod(file_name, 0o440)
+        mongodb_user = pwd.getpwnam(MONGO_USER)
+        os.chown(file_name, mongodb_user.pw_uid, ROOT_USER_GID)
+
```
Comment (MiaAltieri) [inline src/charm.py:252]: copied from VM charm

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -28,3 +38,45 @@ def test_install_snap_packages_failure(self, _call, snap_cache):
...
+            self.harness.charm.set_secret("unit", "somekey", 1)
+
+        self.harness.charm.set_secret("unit", "somekey", "")
+        assert self.harness.charm.get_secret(scope, "somekey") is None
+
+    @patch("charm.MongosOperatorCharm.get_secret", return_value=None)
+    def test_get_keyfile_contents_no_secret(self, get_secret):
+        """Tests file isn't checked if secret isn't set."""
+        self.assertEqual(self.harness.charm.get_keyfile_contents(), None)
+
```
Comment (MiaAltieri) [inline tests/unit/test_charm.py:74]: copied from VM charm

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            return False
+
+        if not self.charm.db_initialised:
+            logger.info("Deferring %s. db is not initialised.", type(event))
+            event.defer()
+            return False
+
+        return True
+
+    def _on_relation_joined(self, event):
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/config_server_interface.py:69]: nits: return type is missing.  It is added to the next function, so for consistency it is better to add it  here as well

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        self.relation_name = relation_name
+        self.charm = charm
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+        # TODO Future PRs handle scale down
+
+    def _on_relation_changed(self, event):
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/config_server_interface.py:119]: nits: return type is missing. It is added to the next function, so for consistency it is better to add it here as well

File: src/charm.py [hunk 5]
```diff
@@ -90,6 +105,151 @@ def _install_snap_packages(self, packages: List[str]) -> None:
...
+    def get_keyfile_contents(self) -> str:
+        """Retrieves the contents of the keyfile on host machine."""
+        # wait for keyFile to be created by leader unit
+        if not self.get_secret(APP_SCOPE, Config.Secrets.SECRET_KEYFILE_NAME):
+            logger.debug("waiting for leader unit to generate keyfile contents")
+            return
+
+        key_file_path = f"{Config.MONGOD_CONF_DIR}/{KEY_FILE}"
+        key_file = Path(key_file_path)
+        print(key_file.is_file())
```
Comment (dmitry-ratushnyy) [inline src/charm.py:227]: looks like debug print

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 3 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/config_server_interface.py:69]: this library was copied over from the VM charm. I can make these changes next time I update that charm (probably today/tomorrow)

File: src/charm.py (hunk 5 above)
Comment (MiaAltieri) [reply src/charm.py:227]: i've been caught :hide-emoji:

File: src/charm.py [hunk 6]
```diff
@@ -90,6 +105,151 @@ def _install_snap_packages(self, packages: List[str]) -> None:
...
+
+        Returns:
+            a list of IP address associated with MongoDB application.
+        """
+        peer_addresses = []
+        if self._peers:
+            peer_addresses = [self._unit_ip(unit) for unit in self._peers.units]
+
+        self_address = self._unit_ip(self.unit)
+        addresses = []
```
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py:154]: 🧠 excellent suggestion, thank you

File: src/charm.py [hunk 7]
```diff
@@ -90,6 +105,151 @@ def _install_snap_packages(self, packages: List[str]) -> None:
                 )
                 raise
 
+    @property
+    def mongos_config(self) -> MongoDBConfiguration:
+        """Generates a MongoDBConfiguration object for mongos in the deployment of MongoDB."""
+        return self._get_mongos_config_for_user(OperatorUser, set(self._unit_ips))
```
Comment (delgod) [inline src/charm.py:111]: in the case of subordinate charm, we should connect only to local mongos
```suggestion
        return self._get_mongos_config_for_user(OperatorUser, {"/tmp/mongos.sock"})
```

mongos should be bond to socket `mongos --bind_ip /tmp/mongos.sock`

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 7 above)
Comment (MiaAltieri) [reply src/charm.py]: At the risk of sounding naive, why not `mongos --bind_ip localhost`

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MiaAltieri) [reply src/charm.py]: actually this function is no longer needed

**Review by delgod — APPROVED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Pull request merged**